### PR TITLE
deck building

### PR DESCRIPTION
**Pull request by notfaj**

### CODE REVIEW

File: wizwalker/memory/memory_objects/window.py [hunk 1]
```diff
@@ -346,7 +348,7 @@ async def card_size_horizontal(self) -> int:
     async def card_size_vertical(self) -> int:
         return await self.read_value_from_offset(0x2A8, "unsigned int")
 
-    async def card_spacing(self) -> int:
+    async def card_spacing_horizontal_adjust(self) -> int:
```
Comment (SirOlaf) [inline wizwalker/memory/memory_objects/window.py]: Why

File: wizwalker/extensions/scripting/deck_builder.py [hunk 2]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
+            number_of_copies_count += 1
+            if number_of_copies:
+                if number_of_copies_count >= number_of_copies:
+                    break
+
+
+    async def change_all_item_card(self):
+        all_cards = await self.get_item_card_names()
+        cards = [*set(all_cards)]
+        for card in cards:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
        for card in all_cards:
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 3]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
+                    pass
+                except AttributeError:
+                    pass
+
+        return list_of_name
+
+
+    async def item_remove_add_by_name(self, name: str, number_of_copies: Optional[int]):
+        ItemSpells_window = await _maybe_get_named_window(self.client.root_window, "ItemSpells")
+        DeckListControl = DynamicDeckListControl(self.client.hook_handler, await ItemSpells_window.read_base_address())
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Bad naming style, use stuff like `item_spells_window` (applies to entire pr)

File: wizwalker/memory/memory_objects/window.py (hunk 1 above)
Comment (notfaj) [reply wizwalker/memory/memory_objects/window.py]: looked at the offsets in CE and it made more sense of it being the horizontal card spacing, I also compared the DeckListControl and SpellListControl as they are similar in some aspects.

File: wizwalker/extensions/scripting/deck_builder.py [hunk 4]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
+    async def get_itemDeckList_card_named(self, name: str) -> list[DeckListControlSpellEntry]:
+        if not self._deck_config_window:
+            self._deck_config_window = await _maybe_get_named_window(self.client.root_window, "DeckConfiguration")
+        ItemSpells_window = await _maybe_get_named_window(self._deck_config_window , "ItemSpells")
+        DeckListControl = DynamicDeckListControl(self.client.hook_handler, await ItemSpells_window.read_base_address())
+        list_of_spells = await DeckListControl.spell_entries()
+        list_of_name = []
+        for spell in list_of_spells:
+            graphical = await spell.graphical_spell()
+            if graphical:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
            if graphical := await spell.graphical_spell():
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 5]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
+        if not self._deck_config_window:
+            self._deck_config_window = await _maybe_get_named_window(self.client.root_window, "DeckConfiguration")
+        ItemSpells_window = await _maybe_get_named_window(self._deck_config_window , "ItemSpells")
+
+        DeckListControl = DynamicDeckListControl(self.client.hook_handler, await ItemSpells_window.read_base_address())
+        list_of_spells = await DeckListControl.spell_entries()
+        list_of_name = []
+        for spell in list_of_spells:
+            graphical = await spell.graphical_spell()
+            if graphical:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
            if graphical := await spell.graphical_spell():
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 6]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
-    async def set_deck_preset(self, preset: dict, *, ignore_failures: bool = False):
-        pass
+        def dict_maker(_list: list):
+            d = {}
+
+            def checkKey(dic: dict, key: str):
+                if key in dic.keys():
+                    return True
+                else:
+                    return False
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion

```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 7]
```diff
@@ -222,7 +699,149 @@ async def get_deck_preset(self) -> dict:
...
+
+            for card in _list:
+                if checkKey(d, card):
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
                if card in d:
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 8]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+            #close and reopen
+            await self.close_deck_page()
+            await self.open_deck_page()
+
+
+    async def click_tc_button(self):
+        TreasureCardButton = await _maybe_get_named_window(self._deck_config_window, "TreasureCardButton")
+        async with self.client.mouse_handler:
+            await self.client.mouse_handler.click_window(TreasureCardButton)
+        await asyncio.sleep(0.5)
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Don't use these hardcoded sleeps, at least let user set the amount

File: wizwalker/extensions/scripting/deck_builder.py [hunk 9]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+        async with self.client.mouse_handler:
+            await self.client.mouse_handler.click_window(leftButton)
+
+        #close and reopen
+        await self.close_deck_page()
+        await self.open_deck_page()
+
+    async def clear_item_deck(self):
+        #item card stuff
         pass
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Why is this empty

File: wizwalker/extensions/scripting/deck_builder.py [hunk 10]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+            y = 266
+        elif card_number == 4:
+            x = 154
+            y = 266
+        elif card_number == 5:
+            x = 59
+            y = 355
+        elif card_number == 6:
+            x = 185
+            y = 355
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: We are not keeping these calculations

File: wizwalker/extensions/scripting/deck_builder.py [hunk 11]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+    async def get_spell_card_names(self) -> list[str]:
+            if not self._deck_config_window:
+                self._deck_config_window = await _maybe_get_named_window(self.client.root_window, "DeckConfiguration")
+            SpellList_window = await _maybe_get_named_window(self.client.root_window, "SpellList")
+            SpellListControl = DynamicSpellListControl(self.client.hook_handler, await SpellList_window.read_base_address())
+            list_of_spells = await SpellListControl.spell_entries()
+            list_of_name = []
+            for spell in list_of_spells:
+                graphical = await spell.graphical_spell()
+                if graphical:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
                if graphical := await spell.graphical_spell():
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 12]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+                        display_name = await template.display_name()
+                        name = await self.client.cache_handler.get_langcode_name(display_name)
+                    except MemoryReadError:
+                        pass
+                    except AttributeError:
+                        pass
+                    except ValueError:
+                        pass
+                    else:
+                        list_of_name.append(name)
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Don't need an else branch for this, put it right after `name = ...`

File: wizwalker/extensions/scripting/deck_builder.py [hunk 13]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+                    try:
+                        template = await graphical.spell_template()
+                        display_name = await template.display_name()
+                        name = await self.client.cache_handler.get_langcode_name(display_name)
+                    except MemoryReadError:
+                        pass
+                    except AttributeError:
+                        pass
+                    except ValueError:
+                        pass
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
                    except (MemoryReadError, AttributeError, ValueError):
                        pass
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 14]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+                    display_name = await template.display_name()
+                    name = await self.client.cache_handler.get_langcode_name(display_name)
+                except MemoryReadError:
+                    pass
+                except AttributeError:
+                    pass
+                except ValueError:
+                    pass
+                else:
+                    list_of_name.append(name)
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Move this up

File: wizwalker/extensions/scripting/deck_builder.py [hunk 15]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+                        if not "tc" in (await template.name()).lower():
+                            break
+                    display_name = await template.display_name()
+                    name = await self.client.cache_handler.get_langcode_name(display_name)
+                except MemoryReadError:
+                    pass
+                except AttributeError:
+                    pass
+                except ValueError:
+                    pass
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
                except (MemoryReadError, AttributeError, ValueError):
                    pass
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 16]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+        DeckListControl = DynamicDeckListControl(self.client.hook_handler, await CardsInDeck_window.read_base_address())
+        list_of_spells = await DeckListControl.spell_entries()
+        list_of_name = []
+        for spell in list_of_spells:
+            graphical = await spell.graphical_spell()
+            if graphical:
+                try:
+                    template = await graphical.spell_template()
+                    if tc:
+                        if not "tc" in (await template.name()).lower():
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
                        if not await template.treasure():
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 17]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+        if not self._deck_config_window:
+            self._deck_config_window = await _maybe_get_named_window(self.client.root_window, "DeckConfiguration")
+
+        CardsInDeck_window = await _maybe_get_named_window(self.client.root_window, "CardsInDeck")
+        DeckListControl = DynamicDeckListControl(self.client.hook_handler, await CardsInDeck_window.read_base_address())
+        list_of_spells = await DeckListControl.spell_entries()
+        list_of_name = []
+        for spell in list_of_spells:
+            graphical = await spell.graphical_spell()
+            if graphical:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]:
```suggestion
            if graphical := await spell.graphical_spell():
```

File: wizwalker/extensions/scripting/deck_builder.py [hunk 18]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+
+    async def get_deck_card_names(self, tc: bool) -> list[str]:
+        """
+        get deck card names
+        
+        Args:
+                tc: True if your reading deck tc's
+        """
+        if not self._deck_config_window:
+            self._deck_config_window = await _maybe_get_named_window(self.client.root_window, "DeckConfiguration")
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: Pull this into its own function, you keep putting this check+fetch

File: wizwalker/extensions/scripting/deck_builder.py [hunk 19]
```diff
@@ -189,30 +205,491 @@ def calculate_icon_position(
...
+
+
+    async def deck_card_number_of_copies(self, card_name: str) -> int:
+        cards_names = await self.get_deck_card_names(False)
+        cards_lower = [c.lower() for c in cards_names]
+        duplicate_cards = [i for i in cards_lower if i == card_name.lower()]
+        return len(duplicate_cards)
+
+
+    async def get_item_card_index(self, card_name: str) -> list[int]:
```
Comment (SirOlaf) [inline wizwalker/extensions/scripting/deck_builder.py]: What if I got multiple of the same tc and want to use only a packed variant for example

**Review by SirOlaf — CHANGES_REQUESTED** [review]
There's more to talk about, but I'll let you address this first

File: wizwalker/extensions/scripting/deck_builder.py (hunk 9 above)
Comment (notfaj) [reply wizwalker/extensions/scripting/deck_builder.py]: I couldn't find a way to differentiate when the items cards where crossed out or not crossed out. So I couldn't tell if the spell was added to the deck, or not added to the deck.

File: wizwalker/memory/memory_objects/window.py (hunk 1 above)
Comment (SirOlaf) [reply wizwalker/memory/memory_objects/window.py]: The same function name seems to appear futher down as well though

File: wizwalker/extensions/scripting/deck_builder.py (hunk 10 above)
Comment (notfaj) [reply wizwalker/extensions/scripting/deck_builder.py]: calculate_card_position and calc_icon_pos?

**Pull request closed without merging**